src/clj/rems/application/expirer_bot.clj

```diff
@@ -5,101 +5,199 @@
+    (when-let [expiration-time (or (get-expiration-time application)
+                                   (calculate-expiration-time expiration application))]
+      (when (time/before? expiration-time now)
+        {:type :application.command/delete
+         :time (time/now)
+         :actor bot-userid
+         :application-id (:application/id application)}))))
+
+(deftest test-expire-application
+  (let [now (time/now)
+        over-90d-ago (time/minus now (time/days 90) (time/seconds 2))
+        over-7d-ago (time/minus now (time/days 7) (time/seconds 2))
+        over-1d-ago (time/minus now (time/days 1) (time/seconds 2))]
+    (testing "without reminders"
+      (is (some? (expire-application {:delete-after "P90D"}
+                                     {:application/applicant {:userid "alice"}
+                                      :application/state :application.state/draft
+                                      :application/last-activity over-90d-ago
+                                      :application/events [{:event/type :application.event/created
+                                                            :event/time over-90d-ago
+                                                            :event/actor "alice"}
+                                                           ;; skip because not applying user
+                                                           {:event/type :application.event/remarked
+                                                            :event/time over-1d-ago
+                                                            :event/actor "hannah"}]}
+                                     now)))
+      (is (nil? (expire-application {:delete-after "P90D"}
+                                    {:application/state :application.state/draft
+                                     :application/last-activity over-7d-ago}
+                                    now))))
+
+    (testing "with reminder set"
+      (is (nil? (expire-application {:delete-after "P90D"
+                                     :reminder-before "P7D"}
+                                    {:application/state :application.state/draft
+                                     :application/last-activity over-90d-ago}
+                                    now))
+          "reminder not yet sent")
+
+      (is (nil? (expire-application {:delete-after "P90D"
+                                     :reminder-before "P7D"}
+                                    {:application/state :application.state/draft
+                                     :application/last-activity over-90d-ago
+                                     :application/events [{:type :application.command/send-expiration-notifications
+                                                           :time over-1d-ago
+                                                           :expires-on now}]}
+                                    now))
+          "reminder sent but time not passed")
```

Comment:
this test looks odd, perhaps something unfinished? (`:application/events` contains command)

Reply:
Fixed 👍